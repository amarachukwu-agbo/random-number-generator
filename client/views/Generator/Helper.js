import FileSaver from 'file-saver';

class Helper {
  static saveNumbersToFile = (phoneNumbers, batch) => {
    const numbersTOCSV = phoneNumbers.join(',').replace(/,/g, '\n');
    const blob = new Blob([numbersTOCSV], { type: 'text/plain;charset=utf-8' });
    FileSaver.saveAs(blob, `phoneNumbers-${batch}.txt`);
  };
}

export default Helper;
