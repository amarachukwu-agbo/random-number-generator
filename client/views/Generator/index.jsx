import React, { Fragment, useState, useEffect } from 'react';
import NavBar from '../../components/NavBar';
import NumbersGrid from '../../components/NumbersGrid';
import Loader from '../../components/Loader';
import Pagination from '../../components/Pagination';
import Header from '../../components/Header';
import { getPhoneNumbers, getMinMaxNumbers } from '../../services';
import './index.scss';

const Generator = () => {
  const [phoneNumbers, setPhoneNumbers] = useState([]);
  const [minMaxNumbers, setMinMaxNumbers] = useState([]);
  const [meta, setMeta] = useState([]);
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState([]);

  const getNumbers = async (page) => {
    try {
      const { phoneNumbers: numbers, meta: metaData } = await getPhoneNumbers(page);
      setPhoneNumbers(numbers);
      setError('');
      setMeta(metaData);
    } catch (err) {
      setError(error);
    } finally {
      setLoading(false);
    }
  };

  const getMinMax = async () => {
    const { minMaxPhoneNumbers } = await getMinMaxNumbers();
    setMinMaxNumbers(minMaxPhoneNumbers);
  };

  useEffect(() => { getNumbers(); }, []);

  useEffect(() => { getMinMax(); }, []);

  return (
    <Fragment>
      <NavBar />
      <div className="generator__container">
        {
        loading ? (<div className="loader"><Loader /></div>) : (
          <Fragment>
            <Header meta={meta} minMaxNumbers={minMaxNumbers} />
            <NumbersGrid phoneNumbers={phoneNumbers} />
            <Pagination metaData={meta} changePage={getNumbers} />
          </Fragment>)
      }
      </div>
    </Fragment>
  );
};

export default Generator;
