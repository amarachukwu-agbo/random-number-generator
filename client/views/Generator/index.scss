body {
  background: #AA076B;  /* fallback for old browsers */
  background: -webkit-linear-gradient(to right, #61045F, #AA076B);  /* Chrome 10-25, Safari 5.1-6 */
  background: linear-gradient(45deg, #AA076B, #61045F); /* W3C, IE 10+/ Edge, Firefox 16+, Chrome 26+, Opera 12+, Safari 7+ */
  height: 100vh;
  margin: 0;
}

.generator {
  &__container {
  padding: 0.7% 4%;
  }
  &__button-group {
    display: flex;
    justify-content: space-between;
    align-items: center;
  }
};

.loader {
  position: fixed;
  top: 40%;
  left: calc(50% - 62.5px)
}
