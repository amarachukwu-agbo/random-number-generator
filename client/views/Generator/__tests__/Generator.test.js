import React from 'react';
import {
  render, fireEvent, cleanup, waitForElement,
} from 'react-testing-library';
import 'jest-dom/extend-expect';
import axiosMock from 'axios';
import Generator from '..';

const mockPhoneNumbers = {
  phoneNumbers: [
    '0197526412',
    '0676706507',
    '0846205003',
    '0745958059',
    '0431347921',
    '0756598026',
    '0171369808',
    '0394491530',
    '0135872917',
    '0241633341',
  ],
  meta: {
    totalCount: 100,
    currentPage: 2,
    pagesCount: 10,
    limit: 10,
  },
  minMaxPhoneNumbers: {
    minNumber: '0135872917',
    maxNumber: '0846205003',
  },
  batchIDs: ['15687', '16743'],
};

const mockGeneratedNumbers = {
  ...mockPhoneNumbers,
  generatedNumbers: mockPhoneNumbers.phoneNumbers,
};
describe('Number Generator', () => {
  beforeEach(() => {
    axiosMock.get.mockImplementation(() => (Promise
      .resolve({ data: mockPhoneNumbers })));
    axiosMock.post.mockImplementation(() => Promise.resolve(
      { data: mockGeneratedNumbers },
    ));
  });

  afterEach(cleanup);

  test('it renders correctly', async () => {
    const { container } = render(<Generator />);
    expect(container.firstChild).toMatchSnapshot();
    expect(axiosMock.get).toHaveBeenCalledTimes(2);
    expect(axiosMock.get.mock.calls[0][0]).toContain('?page=1&sort=ASC');
  });

  test('it fetches the next page of numbers when the next button is clicked', async () => {
    const { getByText } = render(<Generator />);
    const nextButton = await waitForElement(() => getByText('Next'));
    fireEvent.click(nextButton);
    expect(axiosMock.get).toHaveBeenCalledTimes(3);
    expect(axiosMock.get.mock.calls[2][0]).toContain('?page=3&sort=ASC');
  });

  test('it fetches the previous page of numbers when the previous button is clicked', async () => {
    const { getByText } = render(<Generator />);
    const prevButton = await waitForElement(() => getByText('Previous'));
    fireEvent.click(prevButton);
    expect(axiosMock.get).toHaveBeenCalledTimes(3);
    expect(axiosMock.get.mock.calls[2][0]).toContain('?page=1&sort=ASC');
  });

  test('it fetches the min and max numbers', async () => {
    const { getByTestId } = render(<Generator />);
    const maxNumber = await waitForElement(() => getByTestId('Max'));
    const minNumber = await waitForElement(() => getByTestId('Min'));
    expect(axiosMock.get).toHaveBeenCalledTimes(2);
    expect(minNumber.textContent).toBe('0135872917');
    expect(maxNumber.textContent).toBe('0846205003');
  });

  test('it fetches numbers in a batch when the batch is selected', async () => {
    const { getByTestId } = render(<Generator />);
    const dropdown = await waitForElement(() => getByTestId('dropdown'));
    fireEvent.click(dropdown);
    const dropdownItem = await waitForElement(() => getByTestId('dropdown-item-1'));
    fireEvent.click(dropdownItem);
    expect(axiosMock.get).toHaveBeenCalledTimes(3);
    expect(axiosMock.get.mock.calls[2][0]).toContain('numbers/16743?page=1&sort=ASC');
  });

  test('it generates new numbers when the `generate` button is clicked', async () => {
    const { getByText } = render(<Generator />);
    const generateButton = await waitForElement(() => getByText('Generate'));
    fireEvent.click(generateButton);
    expect(axiosMock.post).toHaveBeenCalledTimes(1);
    expect(axiosMock.post.mock.calls[0][0]).toContain('numbers');
  });

  test('it sort numbers in descending order when down sort button', async () => {
    const { getByTestId } = render(<Generator />);
    const sortDownButton = await waitForElement(() => getByTestId('down-button'));
    fireEvent.click(sortDownButton);
    expect(sortDownButton).toHaveClass('active');
    expect(axiosMock.get).toHaveBeenCalledTimes(3);
    expect(axiosMock.get.mock.calls[2][0]).toContain('?page=1&sort=DESC');
  });

  test('it sort numbers in descending order when up sort button', async () => {
    const { getByTestId } = render(<Generator />);
    const sortUpButton = await waitForElement(() => getByTestId('up-button'));
    fireEvent.click(sortUpButton);
    expect(sortUpButton).toHaveClass('active');
    expect(axiosMock.get).toHaveBeenCalledTimes(3);
    expect(axiosMock.get.mock.calls[2][0]).toContain('?page=1&sort=ASC');
  });

  test('it sets error when there is an error fetching numbers', async () => {
    axiosMock.get.mockImplementation(() => Promise.reject(new Error('An error occured')));
    render(<Generator />);
    expect(axiosMock.get).toHaveBeenCalledTimes(2);
  });
});
