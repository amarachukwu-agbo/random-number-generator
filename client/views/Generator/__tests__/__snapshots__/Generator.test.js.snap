// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Number Generator it renders correctly 1`] = `
<div
  class="navbar__container"
>
  <h2
    class="navbar__header--text"
  >
    Phone Number Generator
  </h2>
  <button
    class="numbers__button"
    type="button"
  >
    <span
      class="numbers__button--text"
    >
      Generate
    </span>
  </button>
</div>
`;
